In [2]:
import numpy as np
import matplotlib.pyplot as plt
import random as rand

# Random Walk

See Module 9.5 in *Introduction to Computational Science* by Shiflet and Shiflet.

Many physical processes are a result of an object traveling in a medium through random collisions with molecules or atoms in the medium. Examples include a pollen particle in a fluid, calcium traveling through a bone, or a photon generated in the interior of the sun traveling to the surface. One way to simulate this process is a *random walk*.

In this notebook, you will:

1. randomly select steps for the object.
1. visualize the path of the object.
1. calculate the distance traveled from its starting location.

## Background

Sir Ronald Ross (1857-1932) was a British medical doctor who received the 1902 Nobel Prize in Medicine for his work on malaria. He was not a mathematician by trade, but he was interested in calculating the distance of a mosquito-carried disease from a breeding pool. In September 1904, at the International Congress of Arts and Science in St. Louis, Ross gave a presentation on how to mathematically model the spread of mosquitoes from such a breeding pool.

Introducing the topic, he says:

"Suppose that a mosquito is born at a given point, and that during its life wanders about, to and fro, to left or to right, where it wills, in search of food or mating, over a country which is uniformly attractive to it. After a time, it will die. What are the probabilities that its dead body will be found at a given distance from its birthplace? That is really the problem which governs the whole of this great subject of the prophylaxis of malaria."

He also says:

"The answer depends upon the distance which a mosquito can traverse, not during a single flight, but during its whole life; and so upon certain laws of probability, which must govern its wanderings to and fro upon the face of the earth"

Ross considered the mosquito's life to be divided into $n$ stages. During each stage, it travels a total distance $L$ in the air. He considered mosquitos traveling in all directions, but for mathematical simplicity he considered a single mosquito to travel along one axis only (i.e. one dimension). During each stage, the mosquito randomly decides *twice* to fly forward or backward a distance $L/2$, with equal probability 1/2 of flying forward or backward.

Ross did not give this process a name; however, today mathematicians and scientists call this a *one-dimensional random walk*.

Ross calculated the probability of finding a dead mosquito at a distance $L$, $2L$, $3L$, etc. for $n$ stages. As a specfic example, he considered 1024 mosquitos with $n=5$ stages. The expected number of mosquitos at each possible distance from the breeding pool is shown below.

Distance from the breeding pool | Number of dead mosquitos
:---: | :---:
0 | 252
L | 420
2L | 240
3L | 90
4L | 20
5L | 2
    


## Exercise 1

1. Use Python to calculate the probability of finding a dead mosquito at each of the distances in the table above. 
2. Copy and paste the markdown for the above table, and add a third column that shows the probability you calculated for each distance.

In [19]:
Nmos = 1024

N0 = 252
N1 = 420
N2 = 240
N3 = 90
N4 = 20
N5 = 2

p0 = N0/Nmos * 100
p1 = N1/Nmos * 100
p2 = N2/Nmos * 100
p3 = N3/Nmos * 100
p4 = N4/Nmos * 100
p5 = N5/Nmos * 100

print("At d=0: {:.2f}%".format(p0))
print("At d=1L: {:.2f}%".format(p1))
print("At d=2L: {:.2f}%".format(p2))
print("At d=3L: {:.2f}%".format(p3))
print("At d=4L: {:.2f}%".format(p4))
print("At d=5L: {:.2f}%".format(p5))

At d=0: 24.61%
At d=1L: 41.02%
At d=2L: 23.44%
At d=3L: 8.79%
At d=4L: 1.95%
At d=5L: 0.20%


The probability of finding the mosquito at these distances is shown in the table below.

Distance from the breeding pool | Number of dead mosquitos | Probability (%)
:---: | :---: | :---:
0 | 252 | 24.6
L | 420 | 41.0
2L | 240 | 23.4
3L | 90 | 8.8
4L | 20 | 2.0
5L | 2 | 0.2
    


## Mosquito Model

Let's define a Cartesian coordinate system with $+x$ to the right and $+y$. An object can only be at a location on the grid given by integers $(x,y)$. In other words, the location of the object given by the coordinate pair could be something like $(0,1)$ or $(-10,15)$ but not $(3.5,1)$.  A movement from one location to another is called a *displacement*.

- The mosquito will fly along the x-axis at $y=0$.
- In each stage, a mosquito takes two random "steps", of distance $L/2$ to the right or to the left, with equal probability.
- A step to the right is $\Delta x = +L/2$.
- A step to the left is $\Delta x = -L/2$.
- Use $L=1$ for simplicity.

In the rest of the notebook, we will create a grid and move the mosquito randomly along the x-axis.


## One Mosquito

- Randomly pick 0 or 1 and use this to step right or left.
- Record the x position after each step.

In [23]:
N = 5 #total number of stages
L = 1 #length of stage

#starting position
x = 0

print(f"Stage 0, x = {x}")

#loop
for i in range(N):
    
    #step right or left
    r = rand.randint(0,1)
    if r == 0:
        dx1 = L/2 #step right
    else:
        dx1 = - L/2 #step left

    #step right or left
    r = rand.randint(0,1)
    if r == 0:
        dx2 = L/2 #step right
    else:
        dx2 = - L/2 #step left
    
    dx = dx1 + dx2 #total displacement
    x = x + dx #update the value of x

    print("Stage {:d}, x = {:.1f}".format(i+1,x))


Stage 0, x = 0
Stage 1, x = 0.0
Stage 2, x = 0.0
Stage 3, x = 1.0
Stage 4, x = 0.0
Stage 5, x = -1.0


## Using Python Functions

Use a Python function to do a random walk and return the final x position. I took the previous program, removed the `print` statements, and put it into a function. The variable `N` is passed to the function.


In [24]:
def randomwalk1D(N): #input N, the number of stages; returns final position
    L = 1 #length of stage

    #starting position
    x = 0

    #loop
    for i in range(N):

        #step right or left
        r = rand.randint(0,1)
        if r == 0:
            dx1 = L/2 #step right
        else:
            dx1 = - L/2 #step left

        #step right or left
        r = rand.randint(0,1)
        if r == 0:
            dx2 = L/2 #step right
        else:
            dx2 = - L/2 #step left

        dx = dx1 + dx2 #total displacement
        x = x + dx #update the value of x

    return x
    

To get the final x position of a mosquito, call the function and pass it the number of stages.

In [27]:
Nstages = 5

x=randomwalk1D(Nstages)
print(f"final x = {x}")

final x = 2.0


## Modeling Many Mosquitos

We want to model the *random walk* of 1024 mosquitos. We need a loop to run 1024 times. Create a loop below and call the function `randomwalk1D`. Run the loop 1024 times. Note that when testing, run the loop fewer times (maybe 10 times) and print each final x-position.

In [28]:
Nstages = 5
Nmos = 10 #number of mosquitos

for i in range(Nmos):
    x=randomwalk1D(Nstages)
    print(f"final x = {x}")


final x = 1.0
final x = 0.0
final x = -1.0
final x = 2.0
final x = -1.0
final x = 1.0
final x = 0.0
final x = 1.0
final x = -1.0
final x = 3.0


1. After you know your program is working, copy and paste it into the cell below. 

2. Create variables such as `N0`, `N1`, `N2`, `N3`, `N4`, `N5` that will be used to count the number of times a mosquito dies at a certain distance $0$, $L$, $2L$, etc. 

3. Add code *inside the loop* to:

  - calculate the distance of the final x position of the mosquito. Use `np.abs(x)` to calculate the absolute value of the final x position of the mosquito. 
  - count the number of times each distance $0$, $L$, $2L$, $3L$... occurs


In [10]:
Nstages = 5
Nmos = 10 #number of mosquitos

#variables to count outcomes for distance
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

for i in range(Nmos):
    x = randomwalk1D(Nstages)
    d = np.abs(x) #distance from x=0
    
    if d==0:
        N0 = N0 + 1
    elif d==1:
        N1 = N1 + 1
    elif d==2:
        N2 = N2 + 1
    elif d==3:
        N3 = N3 + 1
    elif d==4:
        N4 = N4 + 1
    elif d==5:
        N5 = N5 + 1
    
    print(f"final distance = {d}")

final distance = 2.0
final distance = 3.0
final distance = 2.0
final distance = 0.0
final distance = 1.0
final distance = 0.0
final distance = 0.0
final distance = 1.0
final distance = 2.0
final distance = 1.0


Copy and paste your program into the cell below. Then, after the loop, calculate the percentage of mosquitos that die at each distance--this is the probability--and compare your results to the table given by Ross.

In [36]:
Nstages = 5
Nmos = 1024 #number of mosquitos

#variables to count outcomes for distance
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

for i in range(Nmos):
    x = randomwalk1D(Nstages)
    d = np.abs(x) #distance from x=0 is |x|
    
    if d==0:
        N0 = N0 + 1
    elif d==1:
        N1 = N1 + 1
    elif d==2:
        N2 = N2 + 1
    elif d==3:
        N3 = N3 + 1
    elif d==4:
        N4 = N4 + 1
    elif d==5:
        N5 = N5 + 1
        
p0 = N0/Nmos * 100 #percentage of mosquitos that die at x = 0
p1 = N1/Nmos * 100 #percentage of mosquitos that die at x = +-1
p2 = N2/Nmos * 100 #percentage of mosquitos that die at x = +-2
p3 = N3/Nmos * 100 #percentage of mosquitos that die at x = +-3
p4 = N4/Nmos * 100 #percentage of mosquitos that die at x = +-4
p5 = N5/Nmos * 100 #percentage of mosquitos that die at x = +-5

print("At d=0: {:.2f}%".format(p0))
print("At d=1L: {:.2f}%".format(p1))
print("At d=2L: {:.2f}%".format(p2))
print("At d=3L: {:.2f}%".format(p3))
print("At d=4L: {:.2f}%".format(p4))
print("At d=5L: {:.2f}%".format(p5))

At d=0: 25.10%
At d=1L: 41.60%
At d=2L: 23.83%
At d=3L: 7.03%
At d=4L: 2.25%
At d=5L: 0.20%


Here's what Ross found.

Distance from the breeding pool | Number of dead mosquitos | Probability (%)
:---: | :---: | :---:
0 | 252 | 24.6
L | 420 | 41.0
2L | 240 | 23.4
3L | 90 | 8.8
4L | 20 | 2.0
5L | 2 | 0.2
    
Our calculated percentages are close to two significant figures.

## Exercise 2

Run your program modeling 1024 mosquitos multiple times and observe the results each time. 

1. Do the percentages (i.e. probabilities) for each distance change?

2. How can you change your program in order to better estimate the expected percentage of mosquitos at each distance?

Compare your results to the table by Ross.

The percentages change slightly. You mostly notice it for the distance $L$. To better estimate the expected percentage, we should run our program for more mosquitos.

In [40]:
Nstages = 5
Nmos = 100000 #number of mosquitos

#variables to count outcomes for distance
N0 = 0
N1 = 0
N2 = 0
N3 = 0
N4 = 0
N5 = 0

for i in range(Nmos):
    x = randomwalk1D(Nstages)
    d = np.abs(x) #distance from x=0 is |x|
    
    if d==0:
        N0 = N0 + 1
    elif d==1:
        N1 = N1 + 1
    elif d==2:
        N2 = N2 + 1
    elif d==3:
        N3 = N3 + 1
    elif d==4:
        N4 = N4 + 1
    elif d==5:
        N5 = N5 + 1
        
p0 = N0/Nmos * 100 #percentage of mosquitos that die at x = 0
p1 = N1/Nmos * 100 #percentage of mosquitos that die at x = +-1
p2 = N2/Nmos * 100 #percentage of mosquitos that die at x = +-2
p3 = N3/Nmos * 100 #percentage of mosquitos that die at x = +-3
p4 = N4/Nmos * 100 #percentage of mosquitos that die at x = +-4
p5 = N5/Nmos * 100 #percentage of mosquitos that die at x = +-5

print("At d=0: {:.2f}%".format(p0))
print("At d=1L: {:.2f}%".format(p1))
print("At d=2L: {:.2f}%".format(p2))
print("At d=3L: {:.2f}%".format(p3))
print("At d=4L: {:.2f}%".format(p4))
print("At d=5L: {:.2f}%".format(p5))

At d=0: 24.48%
At d=1L: 40.96%
At d=2L: 23.47%
At d=3L: 8.95%
At d=4L: 1.94%
At d=5L: 0.21%


These percentages are similar to nearly three significant figures.